switching artists

amitojsingh opened the pull request:
fix #327

amitojsingh committed:
fix audio functionality

amitojsingh committed:
fix tests

amitojsingh committed:
fix deep scan bugs

amitojsingh committed:
Merge branch 'dev' into fix-functionality

amitojsingh committed:
fix: bugs by uttamdeep kaur

amitojsingh committed:
Merge branch 'fix-functionality' into fix-autoplay

amitojsingh committed:
fix: auto play issue with trackPlayer

amitojsingh committed:
added test suite for AudioTrackDialog

amitojsingh committed:
switching artists

amitojsingh committed:
Merge branch 'dev' into switching-artist

amitojsingh committed:
Refactor AudioPlayer and AudioControlBar for improved track handling

- Removed console.log statements from AudioPlayer for cleaner code.
- Updated track selection logic in AudioPlayer to correctly handle progress saving.
- Added loading indicator in AudioControlBar during track seeking.
- Enhanced styles for progress bar and loading overlay.
- Adjusted setAudioProgress action to remove duration parameter for better clarity.

amitojsingh committed:
Refactor AudioControlBar for improved progress handling

- Simplified track ID and lyrics URL retrieval for clarity.
- Ensured reset function is called after saving audio progress.
- Updated tests to verify reset function call after progress update.

Gauravjeetsingh commented on src/ReaderScreen/components/AudioPlayer/components/AudioControlBar/index.test.jsx:
This needs to be updated to include the sequence parameter.

amitojsingh replied:
That's a good catch. I missed it. Thanks.

Gauravjeetsingh commented on src/ReaderScreen/components/AudioPlayer/utils/getSequenceFromPosition.js:
```diff
@@ -0,0 +1,60 @@
+ * @returns {Promise<number|null>} - Start position in seconds or null if not found
+ */
+const getPositionFromSequence = async (lyricsUrl, sequence) => {
+  try {
+    if (!lyricsUrl || sequence == null) {
+      return null;
+    }
+
+    const lrcData = await fetchLRCData(lyricsUrl);
+    if (!lrcData || !Array.isArray(lrcData) || lrcData.length === 0) {
+      return null;
+    }
```
This code is being repeated in both functions, we can create another function `loadLrc` for this.

amitojsingh replied:
fixed

Gauravjeetsingh commented on src/ReaderScreen/components/AudioPlayer/components/AudioControlBar/index.jsx:
```diff
@@ -160,18 +170,40 @@ const AudioControlBar = ({
-          currentPlaying?.id === audioProgress?.[baniID]?.trackId
-        ) {
-          await seekTo(audioProgress?.[baniID]?.position);
+        setIsSeekLoading(true);
```
Should the seek loading begin, before the `addAndPlayTrack` ?

amitojsingh replied:
fixed

Gauravjeetsingh commented on src/ReaderScreen/components/AudioPlayer/index.jsx:
```diff
@@ -129,6 +130,16 @@ const AudioPlayer = ({ baniID, title, webViewRef }) => {
       setCurrentPlaying(selectedTrack);
+      // Save current sequence before switching artists
+      if (selectedTrack?.lyricsUrl && progress?.position != null) {
+        const currentSequence = await getSequenceFromPosition(
```
If there's an error in `getSequenceFromPosition`, it would not save the progress at all. Should we add a try-catch block here to save progress without sequence?

amitojsingh replied:
Saving progress without a sequence doesn’t make sense in this case.
If an artist’s track has no sequence/lyrics file, we cannot accurately map the saved progress to the correct pauri when the user switches artists.

**Why this is a problem**
If we save the raw seek position (e.g., 60 seconds) for an artist that has no sequence, and the user later switches to a different artist:
- The new artist may not be at the same pauri at 60 seconds.
- Each artist recites at a different speed, so their timings will not match.
- This would cause the app to resume playback incorrectly, misaligning the pauri number and confusing the user.

**Example**
- User listens to Bhai Amarjeet Singh Ji, who does not have a sequence file.
- The user stops at the 5th pauri at 60 seconds.
- If we save the position (60 sec) and the user switches to another artist, the other artist will start at 60 seconds of their own track—
   but they may not be at the 5th pauri at that timestamp.
This creates inconsistency and breaks the expected behavior.

**Correct Behavior**
As per the task requirements:
- If no sequence is available, playback must always start from the beginning, because we cannot reliably restore progress.
- Only artists who have a valid sequence/lyrics file can support accurate progress saving and restoring which most of the artist will have.

Let me know if you still have any questions.

Gauravjeetsingh replied:
Makes sense, thanks for the explanation. 👍

amitojsingh committed:
Enhance AudioControlBar and related utilities for better audio progress management

- Added loading state during track seeking in AudioControlBar.
- Refactored LRC data fetching into a separate function for improved readability and error handling.
- Updated tests to accommodate changes in audio progress action parameters.
